news/_posts/2020-11-30-from-legacy-to-future-architecture-connecting-the-dots.md

```diff
@@ -0,0 +1,308 @@
+
+### Extensions need a reliable interface to the Core
+
+As the Core evolves, its components will inevitably need to be split, renamed, change their methods' signatures, and so on. However, doing this would provoke breaking changes for extensions that rely on the core components' interfaces staying unchanged.
+
+This situation can be assimilated to the [double bind dilemma][double-bind-dilemma], where two conflicting needs prevent the situation from evolving satisfactorily:
+
+- The system is complex, has bugs or needs to accommodate new features, **so components must change**.
+- Extensions rely on the current implementation, **so component mustn't change**.
+
+This has been slowing down PrestaShop's progress for quite a while. But luckily, I think there's a way out. 
+
+Let's revisit the D in [SOLID][solid] for a minute. The [Dependency Inversion Principle][dependency-inversion-principle] states the following:
+
+> 1. High level modules should not depend on low level modules. Both must depend on abstractions.
+> 2. Abstractions should not depend on details. The details must depend on the abstractions.
+
+The solution to this dilemma should actually be quite simple: by **making extensions rely on an API** (abstraction) rather than directly upon the Core, the Core will become free to evolve without introducing breaking changes for extensions. 
+
+The answer is to **progressively isolate the Core, then refactor it**.
+
+
+
+## Isolate the Core behind the Core API
+
+The Core API acts as an [anti-corruption layer][anti-corruption-layer], or bridge, between the application code (Controllers, Services, Extensions...) and Core components. Whenever an application level component needs something from the Core, it accesses it through the Core API. 
+
+Thanks to this, application components will no longer depend on Core components directly. Therefore, if for example a Core component should need to be removed and replaced with another, it would be completely transparent from the application point of view, as all changes would happen "behind the scenes".
+
+[![Decoupled dependencies][decoupled-dependencies]][decoupled-dependencies]
+
+To allow this, all code behind this API must become "private", or internal. This means that **once the API is complete, the [SemVer][semver] compatibility engagement will be limited to API components only.** 
+
+Of course, this doesn't mean developers _will not_ be allowed use Core components directly. PrestaShop being an open system, developers will always be free to shape it in any way they want. However, as this API is put in place, the official compatibility engagement will become limited to the API alone, and while an effort will be made to avoid performing breaking changes, they might happen from time to time.
```

Comment:
```suggestion
Of course, this doesn't mean developers _will not_ be allowed to use Core components directly. PrestaShop being an open system, developers will always be free to shape it in any way they want. However, as this API is put in place, the official compatibility engagement will become limited to the API alone, and while an effort will be made to avoid performing breaking changes, they might happen from time to time.
```